clang/test/SemaSYCL/sycl-restrict.cpp

```diff
@@ -378,6 +378,18 @@ __attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
   kernelFunc(); //#call_kernelFunc // expected-note 3{{called by 'kernel_single_task<fake_kernel, (lambda at}}
 }
 
+// Compiler must expand recursion, no errors expected.
+static constexpr unsigned int getNextPowerOf2(unsigned int n,
+                                              unsigned int k = 1) {
+  return (k >= n) ? k : getNextPowerOf2(n, k * 2);
+}
+
+void test_constexpr_recursion() {
+  kernel_single_task<class ConstexprRecursionKernel>([]() {
+    unsigned int x = getNextPowerOf2(3);
```

Comment:
I don't think this is right.  'x' isn't necessarily evaluated at compile time.  An actually evaluated-at-compiletime (at least in ExprConstant) expression should never reach this point of Sema, since it would have already been evaluated.